Fixes #4426 - Add modal wizard to host creation

ShimShtein opened the pull request:
Each discovered host in thelist gets a modal wizard that will ask for organization, location and hostgroup before redirecting to host edit screen.

ShimShtein committed:
Fixes #4426 - Add modal wizard to host creation

lzap commented:
Thanks, this is a great idea that actually solves several issues we have with the edit form while allowing the flexibility to still use it.

Now, the first problem I have is when I click on Provision button and wait about 5 seconds, the modal reloads somehow:

![image](https://cloud.githubusercontent.com/assets/49752/15709680/49d5e81e-2806-11e6-9977-6f266e244491.png)

lzap commented:
For the record, I love the idea and I want to merge it for sure.

Another issue is when I click on Cancel the whole screen remains dimmed for some time, then it returns to normal.

lzap commented:
Tried both Chrome and Firefox stable.

Made log record of the click and extra load: https://gist.github.com/lzap/ef531c2eaaf9330570f7db82fdce24fa

We will need to cover the new dialog with functional tests, but thats later, let's get it working first.

lzap commented on app/helpers/discovered_hosts_helper.rb:
```diff
@@ -75,4 +75,16 @@ def discovery_status_icon(host)
+      tax_html += hidden_tax
+    end
+    tax_html
+  end
```
Where does this gets called?

ShimShtein replied:
It's a hack into host edit form, I add a hidden input field to it that will send org and loc properties to host#update process, so it will update them too.

ShimShtein replied:
They are disabled in the form, so they are not sent to the server, and we cannot set them.

ShimShtein commented:
Fixed the unwanted reload issue and removed the fade effect from the dialog - it should be more immediate now.

ShimShtein commented:
[test]
Test failure seems unrelated, let's try to rerun.

lzap commented:
Ok great, there is one big issue. When I select a Hostgroup in the modal and click on Provision, I get the form with correct HG set, but OS details are not set. I need to switch it back and forth in order to load up details. Looks like we need to trigger 

Can you also make the modal form a little bit narrower and on the other hand, enlarge all the dropdown boxes to the full width? Also I'd perhaps swap Quick Create and Create host buttons to have the blue one on the right side.

There is also a minor issue when using Quick Create - rendering fails after host finishes provisioining:

```
 | Foreman::Exception (ERF42-4516 [Foreman::Exception]: Unknown action name for success message: edit):
 |   app/controllers/application_controller.rb:243:in `process_success'
 |   /home/lzap/work/foreman_discovery/app/controllers/discovered_hosts_controller.rb:73:in `block in perform_update'
```

It also looks like after quick provisitioning, the puppetmaster/ca fields are not inherited (traditional workflow does work fine).

ShimShtein commented:
Fixed:
1. OS details are shown now
2. Fields are wider
3. Buttons swapped

I have also changed the parameters for `#process_success` so it should give a valid message. Unfortunately on my laptop the provisioning is broken, so couldn't test it completely.

I think that puppetmaster and ca should be fixed in quick provisioning too - it kind of comes with the OS fix. Couldn't test it either.,

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -50,20 +50,36 @@ def destroy
+    setup_host_class_variables(@host)
+    @override_taxonomy = true
+    if params[:quick_submit]
+      perform_update(@host, _('Successfully provisioned %s.') % @host.name)
```
I'd prefer without dot.

lzap commented on app/views/discovered_hosts/_discovered_hosts_list.html.erb:
```diff
@@ -38,8 +38,39 @@
       <td class="hidden-tablet hidden-xs"><%= disc_report_column(host) %></td>
       <td>
+        <!-- Modal -->
+      <div class="modal" id="<%= "myModal-#{host.id}" %>" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">
```
Rename `myModal` to reasonable id.

lzap commented:
It works great, couple if minor comments above.

Now, we have zero integration tests coverage of the provisioning flow. Can you add at least one test for quick provision case? I expect this to become the most used form now.

ShimShtein commented:
Fixed small issues and added tests.

lzap commented:
Javascript issues... [test]

lzap commented on app/views/discovered_hosts/_discovered_hosts_list.html.erb:
```diff
@@ -38,8 +38,39 @@
       <td class="hidden-tablet hidden-xs"><%= host.primary_interface.try(:subnet) %></td>
       <td class="hidden-tablet hidden-xs"><%= disc_report_column(host) %></td>
       <td>
+        <!-- Modal -->
```
The comment is out of indentation.

One more thing, I realized that the modal dialog is being rendered per each individual host. It looks like it's all same for all hosts, can you create just one instance of it and little bit of javascript that sets required things?

ShimShtein replied:
> can you create just one instance of it and little bit of javascript that sets required things?

It's doable, but why? It adds more complexity to the view, but I can't see a real benefit except maybe smaller HTML that gets to the client.
From programming point of view, there is only one place to change if we need changes to the modal.
If the amount of the HTML is an issue, we can offload it completely to AJAX.

lzap replied:
Yeah not big deal, its just I have 100 hosts per page setting and page is quite big.

lzap commented:
Merged as 2e08637f70d4d9941b6527793bb300b6824741a6, thank you! Yay!

lzap commented:
Thanks a lot. Let me show you a trick: https://github.com/lzap/bin-public/blob/master/discover-host

Use the script to create fake discovery host, then you can provision them, reboot orchestration step is not taken as an error (this step will obviously fail), so you can easily test discovery in most of cases.

Now with this knowledge, you are more than welcomed to review other patches in this repo :-)

ShimShtein commented:
Cool script! I will surely endorse it.
I did something similar, but way less effective: I just created an empty `Host::Discovered` entry using console and fixed validation errors until I was able to save it. It's enough to show the host in the list, so I was able to check the basic use cases.
Thanks for the review!

lzap commented:
I need to test this patch with Katello, I suspect it won't work with other fields like Content Source, Environment or CV. Not sure.

lzap commented:
Ok that was not the case.
